yjit/src/codegen.rs

```diff
@@ -4637,17 +4637,28 @@ fn jit_rb_int_lshift(
 
     // Untag the fixnum shift amount
     let shift_amt = comptime_shift.as_isize() >> 1;
-
     if shift_amt > 63 || shift_amt < 0 {
         return false;
     }
 
+    // Fallback to a C call if the shift amount varies
+    if asm.ctx.get_chain_depth() > 0 {
```

Comment:
Looks like this effectively disables the specialized path, with `./miniruby --yjit-dump-disasm --yjit-call-threshold=1 -e '23434 << 3'` generating a call:

```
  # regenerate_branch
  # Block: <main>@-e:1 (chain_depth: 1)
  # reg_temps: 00000011
  # Insn: 0004 opt_ltlt (stack_size: 2)
  # call to Integer#<<
  # RUBY_VM_CHECK_INTS(ec)
  0x104d1803c: ldur w11, [x20, #0x20]
  0x104d18040: tst w11, w11
  <snip>
```

Should be `> 1` probably, and it looks like the defer_compilation is from gen_send_general() and it's unconditional.

Reply:
Ugh, thanks for catching it. https://github.com/ruby/ruby/pull/9426/commits/43edc3419f581962122102fb0d00b0c53938dbed

I don't get the point of this increment by the way. If it's only for the "Double defer!" detection, we could assign one bit from `chain_depth_return_landing` to it instead of overloading `chain_depth`. It seems counter-intuitive when you look at `*_MAX_DEPTH` constants.